src/book/verdant-sundial/software-information-entropy.ipynb

```diff
@@ -0,0 +1,530 @@
+    "    file_names: list[str],\n",
+    ") -> dict[str, float]:\n",
+    "    \"\"\"\n",
+    "    Calculates the entropy of changes in specified files between two commits,\n",
+    "    inspired by Shannon's entropy formula. Normalized relative to the total lines\n",
+    "    of code changes across specified files.\n",
+    "\n",
+    "    Args:\n",
+    "        repo_path (str): The file path to the git repository.\n",
+    "        source_commit (str): The git hash of the source commit.\n",
+    "        target_commit (str): The git hash of the target commit.\n",
+    "        file_names (list[str]): List of file names to calculate entropy for.\n",
+    "\n",
+    "    Returns:\n",
+    "        dict[str, float]: A dictionary mapping file names to their calculated entropy.\n",
+    "\n",
+    "    Application of Entropy Calculation:\n",
+    "        Entropy measures the uncertainty in a given system. Calculating the entropy\n",
+    "        of lines of code (LoC) changed reveals the variability and complexity of\n",
+    "        modifications in each file. Higher entropy values indicate more unpredictable\n",
+    "        changes, helping identify potentially unstable code areas.\n",
+    "\n",
+    "    \"\"\"\n",
+    "    loc_changes = calculate_loc_changes(\n",
+    "        repo_path, source_commit, target_commit, file_names\n",
+    "    )\n",
+    "\n",
+    "    # Calculate total lines of code changes across all specified files\n",
+    "    total_changes = sum(loc_changes.values())\n",
+    "\n",
+    "    # Calculate the entropy for each file, relative to total changes\n",
+    "    entropy_calculation = {\n",
+    "        file_name: (\n",
+    "            -(\n",
+    "                (loc_changes[file_name] / total_changes)\n",
+    "                * math.log2(\n",
+    "                    loc_changes[file_name] / total_changes\n",
+    "                )  # Entropy Calculation\n",
+    "            )\n",
+    "            if loc_changes[file_name] != 0\n",
+    "            and total_changes\n",
+    "            != 0  # Avoid division by zero and ensure valid entropy calculation\n",
+    "            else 0.0\n",
+    "        )\n",
+    "        for file_name in loc_changes  # Iterate over each file in loc_changes dictionary\n",
+    "    }\n",
+    "    # Calculate total entropy\n",
+    "    total_entropy = sum(entropy_calculation.values())\n",
+    "\n",
+    "    # Normalize total entropy to range [0, 1]\n",
+    "    max_entropy = len(loc_changes)\n",
+    "    normalized_total_entropy = total_entropy / max_entropy\n",
+    "\n",
+    "    return normalized_total_entropy"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "4a7e76f8",
+   "metadata": {},
+   "source": [
+    "### Proccess Repository"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "id": "0a269d93",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def process_repository(repo_url: str) -> (float, int):\n",
```

Comment:
This function also looks like it might be useful (for running checks in batch)! Consider adding this to the package and then importing within the notebook to help increase code reuse.